core/rust/src/metadata.rs

```diff
@@ -1,15 +1,147 @@
 use crate::error::{DeserializeError, DeserializeFailure, Key};
 use crate::ordered_hash_map::OrderedHashMap;
-use crate::serialization::{Deserialize, LenEncoding, Serialize, StringEncoding};
+use crate::serialization::{Deserialize, LenEncoding, Serialize, StringEncoding, fit_sz};
 use crate::Int;
 use cbor_event::{de::Deserializer, se::Serializer};
 use derivative::Derivative;
 
+use std::collections::BTreeMap;
 use std::io::{BufRead, Seek, SeekFrom, Write};
 
 pub type TransactionMetadatumLabel = u64;
 
-pub type Metadata = OrderedHashMap<TransactionMetadatumLabel, TransactionMetadatum>;
+/// Collection of TransactionMetadatums indexed by TransactionMetadatumLabels
+/// Handles the extremely rare edge-case of in previous generations allowing
+/// duplicate metadatum labels.
+#[derive(Clone, Debug, Default, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct Metadata {
+    pub entries: Vec<(TransactionMetadatumLabel, TransactionMetadatum)>,
```

Comment:
We don't ever expect to see massively combined metadatas with like 1000+ labels, right?

It should be faster this way for reasonable amounts of labels, and you need to remember the order 100% so something like `OrderedHashMap<TransactionMetadatumLabel, Vec<TransactionMetadatum>>` wouldn't work as you could have something like `100`, `50`, `100` which would turn it into `100`, `100`, `50`.

Reply:
Of course you could do a hybrid approach and only remember it that way in the `MetadataEncodings` if we do need to support massive amounts of labels.

Reply:
As in 1000s of collisions? No.

Also, this PR is related to https://github.com/dcSpark/cardano-multiplatform-lib/issues/209

Reply:
Yeah #209 is in the description but I didn't put "fixes" or anything since it might not cover all the cases of duplicate labels. I might 1000s of metadatum labels in general in a single auxdata. Presumably different ones.